sdk/python/lib/test/test_resource.py

```diff
@@ -262,3 +262,36 @@ def test_parent_and_depends_on_are_the_same_12032():
         "myresource",
         opts=pulumi.ResourceOptions(parent=child),
     )
+
+# Regression test for https://github.com/pulumi/pulumi/issues/12736
+@pulumi.runtime.test
+def test_complex_parent_child_dependencies():
+    mocks.set_mocks(MinimalMocks())
+
+    class A(pulumi.ComponentResource):
+        def __init__(self, name: str, opts=None):
+            super().__init__("my:modules:A", name, {}, opts)
+            self.b = B("a-b", opts=ResourceOptions(parent=self))
+            self.c = C("a-c", opts=ResourceOptions(parent=self.b, depends_on=[self.b]))
+
+    class B(pulumi.ComponentResource):
```

Comment:
Took me a second to absorb this example.
For other reviewers, here are the relationships defined in this:

```mermaid
flowchart LR
  a -->|parent of| ab["a-b"] -->|parent of| bchild["b-child"]
  ab -->|parent of| ac["a-c"] -->|parent of| cchild["c-child"]
  d -->|parent of| dchild["d-child"]
  ab -->|parent of| d --> |depends on| ab
  ac -->|depends on| ab
```

Reply:
I wish there was a clear way to add this diagram to the code!

Reply:
We can use ASCII 😆 

```
                       +--------------------------------+                        
                       |          depends on            |                        
                       |                                |                        
                       |                            +-------+                    
                       |                            |       |                    
                       |               +----------->|b-child|                    
                       |               | child      |       |                    
                       v               |            +-------+                    
                   +-------+           |            +-------+                    
+-------+          |       |-----------+            |       |           +-------+
|       |  child   |  a-b  |----------------child-->|  a-c  |  child    |       |
|   a   |--------->|       |<---depends on----------|       |---------->|c-child|
|       |          |       |-----------+            |       |           |       |
+-------+          +-------+           |            +-------+           +-------+
                       ^               |            +-------+           +-------+
                       |               | child      |       |  child    |       |
                       |               +----------->|   d   |---------->|d-child|
                       |                            |       |           |       |
                       |                            +-------+           +-------+
                       |                                |                        
                       |          depends on            |                        
                       +--------------------------------+                        
```

Reply:
❤️  ship it!